ENH: introduce wedge_labels parameter for `pie`

rcomer opened the pull request:
<!--
Thank you so much for your PR!  To help us review your contribution, please check
out the development guide https://matplotlib.org/devdocs/devel/index.html
-->

## PR summary
<!-- Please describe the pull request, using the questions below as guidance, and link to any relevant issues and PRs:

- Why is this change necessary?
- What problem does it solve?
- What is the reasoning for this implementation?

Additionally, please summarize the changes in the title, for example "Raise ValueError on
non-numeric input to set_xlim" and avoid non-descriptive titles such as "Addresses
issue #8576".
-->
The proposal evolved through the below discussion.  The idea is to introduce a new _wedge_labels_ parameter to `pie` which supersedes *autopct*, via its use of the `str.format` method, but it can also take a list of labels consistent with the existing *labels* parameter.

* Introduce *wedge_labels* as a shortcut to the `pie_label` method introduced in #30733.
* Introduce *wedge_label_distance* (default 0.6) to specify how far the *wedge_labels* are from the pie centre.
* Deprecate the *labeldistance* parameter in two steps: First change the default to `None` (which means *labels* are only used for the legend).  Later remove this parameter altogether - direct to use *wedge_labels* instead. 
   * This differs from the approach agreed below because it gets us there in fewer steps, and gives downstream more overlap when *wedge_labels* is available and *labeldistance=float* is not yet deprecated. See https://github.com/matplotlib/matplotlib/pull/29152#issuecomment-2559382405.
* Discourage using `autopct` in favour of `wedge_labels`.


## PR checklist
<!-- Please mark any checkboxes that do not apply to this PR as [N/A].-->

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [x] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [x] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [x] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

<!--We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.-->

rcomer commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -3275,9 +3295,7 @@ def pie(self, x, explode=None, labels=None, colors=None,
-        # The use of float32 is "historical", but can't be changed without
-        # regenerating the test baselines.
-        x = np.asarray(x, np.float32)
+        x = np.asarray(x)
```
Casting to float here meant that we couldn't use "%d" for the format string, which is the most obvious thing to use if your input values are int.  No tests failed for me locally.  I note that tolerances were recently added for numpy v2:
https://github.com/matplotlib/matplotlib/blob/183b04fb43f57ffe66da68c729a179e397cd35f6/lib/matplotlib/tests/test_axes.py#L5947-L5950

rcomer commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -3369,15 +3389,33 @@ def get_next_color():
+
+            elif absolutefmt is not None:
+                if isinstance(absolutefmt, str):
+                    s = absolutefmt % n
```
It would be nice to support `absolutefmt.format(n)` here as well, and similarly for `autopct`  I vaguely remember seeing we had some logic _somewhere_ that could distinguish between old and new style format strings and apply them appropriately.  Did I dream that?  If not, could someone remind me where I saw it?

rcomer replied:
I did not dream it
https://github.com/matplotlib/matplotlib/blob/183b04fb43f57ffe66da68c729a179e397cd35f6/lib/matplotlib/cbook.py#L2369-L2377

rcomer commented:
[What's new entry](https://output.circle-artifacts.com/output/job/898ef810-222a-4205-b681-690867b7f850/artifacts/0/doc/build/html/users/next_whats_new/pie_abs_labels.html)
[Addition to Pie Charts example](https://output.circle-artifacts.com/output/job/898ef810-222a-4205-b681-690867b7f850/artifacts/0/doc/build/html/gallery/pie_and_polar_charts/pie_features.html#auto-label-slices)
[Simplified Matplotlib Bakery](https://output.circle-artifacts.com/output/job/898ef810-222a-4205-b681-690867b7f850/artifacts/0/doc/build/html/gallery/pie_and_polar_charts/pie_and_donut_labels.html#sphx-glr-gallery-pie-and-polar-charts-pie-and-donut-labels-py)

timhoffm commented:
If we are touching the existing API, should we reconsider more fundamentally?

Currently, `pie()` has two sets of annotations
1. "label", which are strings and by default displayed outside next to the wedges
2. "pct", which are percent numbers and by default displayed inside the wedges

#### Some observations:
- "pct" is a too specific name, "numbers" would be better.
- That labels are outside and pct inside is an arbitrary choice.

#### What we fundamentally need to keep:
The two sets of annotations must stay. We'd loose functionality if we allowed only one label. Also, a generalization to more than two sets of annotations does not make sense.

Side-note: If I was designing this from scratch, I'd likely focus on location `inner_label` and `outer_label`, and each could be (i) an explicit list of strings (ii) a function (iii) a format-string, likely `{}`-based, because that allows direct percent-formatting without needing numeric action on our side. However, I belive this would be too much of a change for the current API.

Therefore, I propose to
- keep `labels`, `label_distance` as is
- introduce `inner_labels` and `inner_label_distance`. `inner_labels` should take (i) a list (ii) a function (iii) a `{}` format string
- `autopct` translates to `inner_labels` and giving both is an error. Discourage `autopct` in favor of `inner_labels` and describe the transition (note that autopct's %-format and callable take values that are scaled by 100).
- `pctdistance` translates to `inner_label_distance` and giving both is an error. Discourage `pctdistance` in favor of `inner_labels_distance` and state that you should use the variable that is matching to the one used to define the labels `inner_labels=..., inner_label_distance=...` and `autopct=..., pctdistance=...`.
- optionally, expand `labels` to also take (ii) a function (iii) a `{}` format string

Does that sound reasonable?

timhoffm commented:
What I've missed in my above thought is that `{}` direct percent-formatting takes case of the factor 100 between relative and percent numbers, but it does not solve the issue where to pass relative (or percent) vs. absolute numbers to the formatter. That extra information that we'd need to get in somehow.

I'm trying to find a consistent way to map the option space:
- We need potentially two labels per wedge, typically one of them is inside and one is outside, but the positions are freely adjustable.
- We have three kinds of information: 1. Fixed arbitray labels 2. absolute numbers 3. percent (or relative numbers)
  Note that the number variants are convenience. A user could take the input values, transform them to whatever string they want and pass them as fixed label.

There are two ways for structuring:
- Have two label mechanisms that are structurally equivalent (i.e. allow fixed values and numerics). The distinction would be the (default) position. This was my thought above.
- Distinguish by semantics, i.e. one set of args for fixed labels, one set of args for numeric labels, ...
  Here we actually get into some complication because either we have to define the sets of args (fixed, absolute, percent) but map them to two positions; or we have to merge absolute and percent to one parameter to one numbers parameter and add a flag to determine absolute vs relative. Either way is messy.

IMHO the "absolute values" request comes from the fact that the inner label is too specific in that it only handles percent. There's no API to get arbitrary text in there.

A minimal green-field approach would be `inner_labels` and `outer_labels` (or `labels` when considering compatibility) as lists of fixed texts. Everything else is convenience. Possibly, we don't even need absolute value support then. It is just `pie(x, inner_labels=[f'{xi:.1f}' for xi in x])` IMHO this is even better - because more explicit - than `pie(x, absolutefmt="%.1f")`. Percent formatting is only marginally more complex `pie(x, inner_labels=[[f'{xi:.1%}' for xi in x/x.sum()])`.

I therefore revise my above proposal to the following minimal change:

- keep `labels`, `label_distance` as is
- introduce `inner_labels` and `inner_label_distance`. `inner_labels` should take a list only
- `autopct` translates to `inner_labels` and giving both is an error. Describe `autopct` as a higher-level convenience function to get percent values into inner_labels.
- Document `pctdistance` as equivalent to `inner_label_distance` and giving both is an error. State that you should use the variable that is matching to the one used to define the labels `inner_labels=..., inner_label_distance=...` and `autopct=..., pctdistance=...`.

One can later decide whether both of the labels parameters should grow some formatting / function capability.

rcomer commented:
I am also wondering if it’s worth factoring out a public `pie_label` method, to rhyme with `bar_label`.  The existing `labels` and `autopct` could remain as conveniences that internally call this method, but a user could also call it directly as many times as they like to get whatever combination of labels they like.

timhoffm commented:
`pie_label` method: I'm a bit sceptical. If we didn't have labelling functionality at all in `pie`, that might be the way to go. But as is, there's not much new functionality to be introduced via a `pie_label`. More than two positions (inside/outside) would be a very edge case. And the fancy number formatting/percent stuff is quite optional, because you can do it out-of-the-box with little overhead I wouldn't introduce new just API for that.

What may be a good idea is a private `_wedge_text()` method as internal refactoring, which would simplify the label handling in pie.

rcomer commented:
OK let me see if I can summarise the current thinking:

* Introduce `inner_labels` and `outer_labels` which take a list of strings (and maybe func and/or "{}" formatting for pie fractions)[^1].
* Introduce `inner_label_distance` (default 0.6) and `outer_label_distance` (default 1.1) to specify how far the above are from the pie centre.
* Deprecate using `labels` with not-None `labeldistance`  - direct to use outer labels instead.[^2]
* Discourage[^3] using `autopct` in favour of `inner_labels`.  Passing both is an error.
* Passing `pctdistance` and `inner_labels` is an error.
* Passing `inner_label_distance` and `autopct` is an error.

[^1]: I think we should offer the formatting option as it will be hard to discourage use of _autopct_ if the replacement is less convenient.
[^2]: when this deprecation has expired, deprecate the `labeldistance` parameter altogether.
[^3]: it might be nice to eventually deprecate `autopct` as having 4 parameters for labels is confusing.

timhoffm commented:
Good summary! This is *one* viable way to make the API more consistent.

> I think we should offer the formatting option as it will be hard to discourage use of autopct if the replacement is less convenient.

formatting is a bit tricky because you have to decide whether the inputs should be absolute values or fractions. You cannot have both (at least not without an extra parameter). I'm inclined to expect absolute values by default, and that would still not give a replacement for autopct. We could decide to go with relative values, but it feels a bit unintuitive `pie(x, inner_labels="%.1f")`.

The function approach is a bit better, because you could support two signatures `def fmt(abs_value: float) -> str` and `def fmt(abs_value: float, total: float) -> str`. The first one is a simpler version for convenince, the second one would support percents as `pie(values, inner_label=lambda x, total: f'{x/total:.1%'})`.

---

### Alternative: pie_label()

I've thought about this a bit more. The idea is fundamentally reasonable. On the plus side, we get better decoupling and full control over each label set - currently `textprops` are shared and `rotatelabels` only applies to the outside.

Problem 1: AFAICS, we cannot retrieve the underlying data value from a wedge. Therefore no external function can create labels based on the output of `pie()`. We'd likely need to create a semantic artist `PieWedge` or a kind of `Pie` container class to handle this.
Problem 2: Removing all labelling capability from `pie` and forcing the use of additional functions feels a bit drastic. The alternatives would be to keep some labelling capability and that's not great either: (i) Keeping `labels` as the simple generic variant keeps the entanglement with legend (ii) Keeping autopct is very specific (iii) changing to a new single labels API forces a migration and makes keeping an additional text labels API only half as valuable.

If feel this is less good than the summarized approach.

jklymak commented:
Completely green fielding this issue a I'm pretty ignorant about `pie` - I would personally get rid of autopcnt - it's not like our users cannot calculate the data in terms of a percentage, and the pie chart would look the same.  Then the formatting of the quantitative label would be straightforward.

rcomer commented:
I would prioritise formatting fractions over absolute values:

- Since plotting fractions is basically the point of `pie` I don’t think it’s too unintuitive that those would have the convenience in terms of labelling.  The canonical example would be `pie(x, inner_labels="{:.1%}")`.  Also since `pie` already calculates the fractions in order to draw the wedges it feels a bit annoying/redundant for the user to have to pre-process the data just for the labels (even if it is a simple and cheap calculation).
- Absolute values are slightly more straightforward to convert to the desired string.  In fact if the user has ints then they may not care about formatting, so they could just pass the list of numbers as labels and the `Text` constructor will take care of converting them to strings.

timhoffm commented:
> Edit: I guess you _could_ detect whether the format string contains "%}". If it does, give it fractions. If not give it absolute values.

Yes, I've briefly thought about that too, but it's too much magic 🪄 . There's a standard expectation how `{.1f}` and `{.1%}` relate to each other. We shouldn't mess with that. Then let's rather do the `percent:...` prefix.

rcomer commented:
Named formats would give a lot more flexibility...

```python
In [2]: '{abs:d}'.format(abs=42, frac=0.8)
Out[2]: '42'

In [3]: '{frac:.1%}'.format(abs=42, frac=0.8)
Out[3]: '80.0%'

In [4]: '{abs:d} ({frac:.0%})'.format(abs=42, frac=0.8)
Out[4]: '42 (80%)'
```

timhoffm commented:
> Named placeholders would give a lot more flexibility...

Great idea. Add this format-string as acceptable input for `inner_labels` and `outer_labels` in https://github.com/matplotlib/matplotlib/pull/29152#issuecomment-2483918655 and we have a compact and clean API.

The fundamental question is: Do we want to go through the hassle of repurposing `labels` to only apply to legend. If yes, this is the reasonable way forward. If no, only do `inner_labels` as a replacement of `autopct`. I'm 0.5 on going all the way (but slowly, i.e. start with a pending deprecation).

timhoffm commented:
Side note on naming: I've briefly pondered whether in would make sense to bikeshed `outer_labels`, `inner_labels` to replace the "label" part; motivation: set it more clearly apart from `labels`; e.g. go for `wedge_text` or similar. However, we have ample precedence for `*_label` usages independent of `labels` mapping to legend: `bar_label`, `clabel` (contour label), `tick_labels`, `xlabel`, `ylabel`. Therefore `outer_labels` and `inner_labels` seem perfectly fine.

timhoffm commented:
> Though could this also be the external function `pie_label`?

Not easily. See https://github.com/matplotlib/matplotlib/pull/29152#issuecomment-2484257688 for the issues I see with  `pie_label` function.

rcomer commented:
What would be the default for `wedge_label_distance`?

timhoffm commented:
Searching for pie chart images on google has significantly more inner labels than outer labels. So inner placement should be the default, exact number t.b.d. (likely 0.6 currently used for pctdistance).

rcomer commented:
I see the attraction of the _wedgelables_ tuple parameter for the reasons @story645 stated.  However I think I still lean to the separate inner/outer labels:

- I think it will be easier to explain/document.
- I worry that we generate a lot of support questions from users who passed a list instead of a tuple and got unexpected results.
- Someone who only wants outer labels can have them with a single parameter.
- As @timhoffm noted above, wanting more than two sets of labels is probably rare.

timhoffm commented:
> I see the attraction of the _wedgelables_ tuple parameter for the reasons @story645 stated. However I think I still lean to the separate inner/outer labels:
> 
> * I think the separate inner/outer labels will be easier to explain/document.
```
wedge_labels : list of str or list of list of str, optional
    Labels attached to the wedges. The radial position of the labels is defined by
    *wedge_label_distance*. You can have one or multiple sets of labels.

    A single list of strings defines one label per wedge. The length must match the
    length of *x*. Example::
     
        wedge_labels=["wedge_1", "wedge_2", "wedge_3"]
     
    You can attach multiple labels per wedge by passing a sequence of lists i.e. ::
     
        wedge_labels=[
            ["wedge_1_label_1", "wedge_2_label_1", "wedge_3_label_1"],
            ["wedge_1_label_2", "wedge_2_label_2", "wedge_3_label_2"],
        ]

wedge_label_distance : float or list of float, default: 0.6
    The radial position of the wedge labels, relative to the pie radius.
    Values > 1 are outside the wedge and values < 1 are inside the wedge.
    
    If you use more than one set of *wedge_labels*, you have to sepecify the
    distance for each set of labels. Example::
    
        wedge_labels=[
            ["wedge_1_inner", "wedge_2_inner", "wedge_3_inner"],
            ["wedge_1_outer", "wedge_2_outer", "wedge_3_outer"],
        ],
        wedge_label_distance=(0.6, 1.3)
```		

> * I worry that _wedgelables_ might generate a lot of support questions from users who passed a list instead of a tuple and got unexpected results.

With a bit of effort, we should be able to *do the expected thing* for any reasonable input. Essentially, we can detect if its `Sequence[str]` or `Sequence[Sequence[str]]`.

> * Someone who only wants outer labels can have them with the single _outer_labels_ parameter.

Admittedly, `wedge_labels=..., wedge_label_distance=1.2` is slightly more complicated than `outer_labels=...`, but only gradually. And given that outer labels are comparably rare, that's bearable.

Overall, I feel both approaches the `wedge_labels` and `inner/outer_labels` are perfectly managable and concise. To me it boils down to a semantic preference: Do we want to encode the inner/outer concept or do we want generic labels that can be placed in differenent positions.

rcomer commented:
OK I am reassured by the above.  I am happy to work on implementing this with _wedgelabels_ if we agree we want to go ahead.

rcomer commented:
> > Do we want to encode the inner/outer concept
> 
> Would we enforce the semantics via value checking? My assumption is that that would break lots of existing code.

I had assumed that users would be free to swap them if they like, and the names only refer to the default positions.

rcomer commented:
Revised proposal:

* Introduce `wedge_labels` which take a list of strings or a format string to be formatted with `.format(abs=abs, frac=frac)`, or a sequence of those.
* Introduce `wedge_label_distance` (default 0.6) to specify how far the above are from the pie centre.
* If `wedge_labels` has multiple sets of labels, `wedge_label_distance` must be a sequence of the same length.  Otherwise error.
* Deprecate using `labels` with not-None `labeldistance`  - direct to use `wedge_labels` instead.[^2]
* Discourage[^3] using `autopct` in favour of `wedge_labels`.  If passed, internally incorporate it into the `wedge_label` handling.

[^2]: when this deprecation has expired, deprecate the `labeldistance` parameter altogether.
[^3]: it might be nice to eventually deprecate `autopct` as having 3 parameters for labels is confusing.

rcomer commented:
> With a bit of effort, we should be able to do the expected thing for any reasonable input. Essentially, we can detect if its `Sequence[str]` or Sequence[Sequence[str]]`.

`Sequence[str]` may either mean one string for each wedge or one format string for each set of labels.  If the user made no mistakes then we can tell which of these it is by whether `wedge_label_distance` is a sequence of matching length.  If the user passes multiple format strings but forgot to pass `wedge_label_distance` then we can error if the number of format strings does not match the number of wedges.  If the number of format strings matches the number of wedges then I think there's nothing we can do and they get unexpected output.  Perhaps this is an unlikely scenario and I am overthinking it...

timhoffm commented:
Sorry, I've not remembered the format-string variant. The cases N wedges with N `wedge_labels` are indeed ambiguous by the structure of `wedge_labels`. Nevertheless, in these special cases we can check how many elements `wedge_label_distance` has. It could still be that the user forgot to specify `wedge_label_distance` and wanted N but used the default 1. If you want to additionally guard against this / don't take `wedge_label_distance` as a criterion, it may be better to check that the strings are format strings*.

* I believe the best check for a format string is `is_fmt_str = s.format(abs=..., frac=...) != s`. Note that the kwargs are not mandatory to be consumed so that a plain string will be simply returned unchanged. Also, "containing curly braces is" not a reliable criterion for a format string and you don't want to build a regex for a valid format string.

rcomer commented:
Switching labels as I believe we’ve settled on https://github.com/matplotlib/matplotlib/pull/29152#issuecomment-2551136479.

rcomer commented:
Is there contributor guidance somewhere (or an example) for documenting a pending deprecation?

rcomer commented:
Belatedly I realised there is some difference in handling between the current `labels` and autolabels.  So it is not as generalisable as I had assumed.  The `labels` have horizontal alignment set to the side closest to the centre.  This makes sense when placing the labels outside the pie so that they do not overlap the wedges.  They also have an option to automatically rotate so they are orientated relative to the circle.
https://github.com/matplotlib/matplotlib/blob/677d99097d4d51c12a564146763d8db9a439c307/lib/matplotlib/axes/_axes.py#L3401-L3407

OTOH the automatic labels are just aligned to the centre.
https://github.com/matplotlib/matplotlib/blob/677d99097d4d51c12a564146763d8db9a439c307/lib/matplotlib/axes/_axes.py#L3430-L3433

Do we want to make horizontalalignment change based on whether `wedge_label_distance` is greater or less than 1?  Or is that too much magic?

Do we want to support the `rotatelabels` option for all labels?  If so does it become a list of bools, similar to now having a list of distances?

Apologies, I should have spotted these things much earlier.

timhoffm commented:
> Apologies, I should have spotted these things much earlier.

Thanks for digging this up, and no apology needed. I also haven't thought about it. These are the things that pop up when going from theory to code.

To the question: The current alignment behavior is largely reasonable. Take this example:
![image](https://github.com/user-attachments/assets/0b25c50a-0b5f-4bdc-b0e5-76580c21abe7)

- The inner labels have to horizontally center. Otherwise, they'd only render into one half-side of the wedge.
- For outer labels have to "attach" to the border of the pie. Centering doesn't work because long labels would spill into the pie. The simple solution is to adjust the horizontal alignment based on the position. This gives reasonable results because text is typically wider than high. If you want to be more to-the-point, the alignment would change depending on the position like this (with t.b.d. angles of transition):
![image](https://github.com/user-attachments/assets/50a8c5e9-5ccf-4439-9e9e-e009878a6e5e)
You can also see from this picture that keeping the vertical alignment always centered does not make much of a difference: The labels only move a bit closer to the wedge by half the font height.

To answer your questions:

> Do we want to make horizontalalignment change based on whether `wedge_label_distance` is greater or less than 1? Or is that too much magic?

Yes. This is necessary to get the expected / reasonably looking behavior.

> Do we want to support the `rotatelabels` option for all labels? If so does it become a list of bools, similar to now having a list of distances?

Probably yes. It's unlikely that one wants to rotate the inner labels, but having the parameter and only apply it if the labels are outside feels a bit weird. (Note: The difference to the alignment is that the distance-dependent alignment is an automatic choice behind the scenes, but `rotatelabels` would be explicitly given by the user).

*Overall remark:* These topics make inner and outer labels less equal, which is an argument in favor of dedicated `inner_labels`/`outer_labels`. I'd be open to revisit that discussion in the light of the new insights. But I believe it's not strong enough to exclude the generic labels version.

rcomer commented:
> > Is there contributor guidance somewhere (or an example) for documenting a pending deprecation?
> 
> https://matplotlib.org/devdocs/devel/api_changes.html
> 
> Please let me know what's missing from there and how we can make it clearer that that's where those docs are

Thanks - I am aware of that page and it's great for how to deprecate something in the next meso release.  However, I think we recently agreed that something should not be deprecated unless the alternative has been available for at least one meso release.  So if we introduce `wedge_labels` in v3.11 we do not deprecate `labeldistance not None` until v3.12.  I assume for the warning I do it as normal but set `pending=True`, but am unclear if I need to document as normal or do something different.

rcomer commented:
> If you want to be more to-the-point, the alignment would change depending on the position like this (with t.b.d. angles of transition)

I like this - it would be good to revisit once I have something working with the simpler current logic.

> These topics make inner and outer labels less equal, which is an argument in favor of dedicated inner_labels/outer_labels. I'd be open to revisit that discussion in the light of the new insights. But I believe it's not strong enough to exclude the generic labels version.

I think either way internally I build a list of lists, so it should be pretty easy to switch later if we decide to go that way.  For now I will proceed with `wedge_labels`.

rcomer committed:
ENH: add wedge_labels parameter for pie

rcomer commented:
Currently the fontsize used for `labels` is `rcParams['xtick.labelsize']` whereas for the auto labels no choice is made so I assume it is `rcParams['font.size']`.  `rcParams['xtick.labelsize']` defaults to 'medium' so if the user didn't touch `rcParams` then I think these will be the same.  These choices predate Matplotlib being in git.  Does anyone have an opinion what these should be for `wedge_labels`?  I note that

* The user can already override the fontsize for all labels using the _textprops_ parameter, though that will make them all the same size.
* The `Text` objects are all returned from the function so the user can update properties on individual labels there, though that will of course be more fiddly.

Obvious options

1. Make them `rcParams['xtick.labelsize']` if distance > 1, otherwise `rcParams['font.size']`.  Advantage: more flexibility for the user to choose different sizes.  Probably consistent with thinking of the pie as a polar axes with the outer labels as tick labels and the inner labels as annotations.
2. Just make them all `rcParams['font.size']`.  Advantage: perhaps the least surprising for the user.
3. Just make them all `rcParams['xtick.labelsize']`.  I'm not seeing an obvious advantage of this.

We could also make _textprops_ more flexible to take a dictionary for each set of labels, either now or as a future enhancement.

rcomer commented:
Still very WIP.  I only pushed so that the branch is backed up!

timhoffm commented:
I’d intuitively go with that

> 2\. Just make them all `rcParams['font.size']`.  Advantage: perhaps the least surprising for the user.

If you want to look for more analogies, check `annotate()`, `bar_label()`, `clabel()`, …

rcomer commented:
Thinking again about the _labeldistance_ parameter, instead of

1. Now: pending deprecation for setting it to anything other than `None`
2. Later: full deprecation for above
3. Even later: deprecate parameter completely as it can now only be `None`

we could do

1. Now: behaviour change - in 2 releases time `None` becomes the default.
2. Later: deprecate parameter; direct to use _wedge_lables_ instead 

This gets us there in one less step and one less release.  Downstream can silence the immediate deprecation warning by just passing `labeldistance=1.1`.  That will work fine with previous versions and also gives two versions of overlap where _wedge_labels_ will also work.

rcomer commented:
There is a [test case](https://github.com/matplotlib/matplotlib/blob/c01eb01da8321032cd4a461cb2236701de876727/lib/matplotlib/tests/test_axes.py#L6020-L6022) that passes `labels="l"` where "l" is a key for the `data` dictionary.  It also passes _autopct_.  I have not tried, but I think `wedge_labels=['l', format_string]` would not work without making changes in the `_process_data` decorator.

Do we still need to support cases that this test represents?  If so, should we go back to `inner_labels`/`outer_labels`?  Or modify `_process_data`?  Or is there another way?

rcomer commented:
For now I have modified `_preprocess_data` to work on a sequence, though I am unclear whether we really want to do that for such a widely used decorator.

I'm also generally a bit concerned about how much new code there is here for a relatively small feature.

rcomer commented:
Thanks @timhoffm.  You didn't let it slip - it was me who had it in draft for most of the year!

Thanks to your suggestion at https://github.com/matplotlib/matplotlib/pull/29152#issuecomment-2551381636, we detect the difference between format strings and plain strings so the ambiguity is properly handled:

```python
plt.pie([5, 2], wedge_labels=["spam", "eggs"])
```
<img width="640" height="480" alt="image" src="https://github.com/user-attachments/assets/10e07427-83af-4fc2-b876-65bf45036115" />

```python
plt.pie([5, 2], wedge_labels=["{absval:d}", "{frac:.0%}"])
```
```
ValueError: Found 2 sets of wedge labels but 1 wedge label distances
```
```python
plt.pie([5, 2], wedge_labels=["{absval:d}", "{frac:.0%}"], wedge_label_distance=[0.6, 1.1])
```
<img width="640" height="480" alt="image" src="https://github.com/user-attachments/assets/065483aa-b084-4fef-a55a-b75e302909d4" />

So I do not think there is any technical problem with using *wedge_labels*, we only need to decide from an API point of view whether it's intuitive enough.  I know @story645 was keen on the single *wedge_labels* parameter.

timhoffm commented:
> So I do not think there is any technical problem with using _wedge_labels_, we only need to decide from an API point of view whether it's intuitive enough.

Exactly. We have the following use cases (Note: i use `fractions` and `names` as placeholders for the labels. These could be either lists or format strings):

1. One inner label:
```python
ax.pie(values, wedge_labels=fractions)
ax.pie(values, inner_labels=fractions)
```

2. One outer label:
```python
ax.pie(values, wedge_labels=fractions, wedge_label_distance=1.1)
ax.pie(values, outer_labels=fractions)
```

3. Inner and outer labels (standard distances):
```python
ax.pie(values, wedge_labels=[fractions, names], wedge_label_distance=[0.6, 1.1])
ax.pie(values, inner_labels=fractions, outer_labels=names)
```
3b. or if we add some magic, the two wedge_labels case could automatically result in default distances
```python
ax.pie(values, wedge_labels=[fractions, names])
ax.pie(values, inner_labels=fractions, outer_labels=names)
```

4. Inner and outer labels (custom distances):
```python
ax.pie(values, wedge_labels=[fractions, names], wedge_label_distance=[0.5, 1.3])
ax.pie(values, inner_labels=fractions, outer_labels=names, inner_label_distance=0.5, outer_label_distance=1.3)
```

Looking at the code, I did a personal subjective usability rating - feel free to do yours.

|                                                | *Relevance (1-5)* | `wedge_labels` | `inner_labels`, `outer_labels` |
|------------------------------------------------|-------------------|----------------|--------------------------------|
| 1. One inner label                             | 5                 | ⭐⭐⭐⭐⭐          | ⭐⭐⭐⭐⭐                          |
| 2. One outer label                             | 5                 | ⭐⭐⭐⭐           | ⭐⭐⭐⭐⭐                          |
| 3. Inner and outer labels (standard distances) | 3                 | ⭐⭐⭐            | ⭐⭐⭐⭐⭐                          |
| 3b. Inner and outer labels (magic defaults)    | 3                 | ⭐⭐⭐            | ⭐⭐⭐⭐⭐                          |
| 4. Inner and outer labels  (custom distances)  | 2                 | ⭐⭐⭐            | ⭐⭐                             |
| 5. More than two labels                        | 1                 | ⭐⭐             | N/A                            |

Overall, `inner_labels`, `outer_labels` perform better on the common use cases, because they are more explicit. They are only slightly worse for custom positioning of two labels (very verbose) and cannot handle more than two labels, wich is a very edge case.

story645 commented:
> inner_labels, outer_labels perform better on the common use cases, because they are more explicit

I don't like them b/c we don't force any semantic constraints on them, which means realistically you can use either one for the other case. Consistently the "why are there 3 ways to do this?" is what creates confusion.

If I'm setting multiple labels, I like being able to use two parameters over 4. Ideally, I'd love a (label, distance) tuple to remove coupled keywords, but my guess is that gets very messy very fast from a parsing pov.

jklymak commented:
I didn't read all of the above, but did you consider ax.label_wedges or some such?  It has a precedent in contour and clabel, and multiple labels could easily be implemented by multiple calls.

timhoffm commented:
I think you'd need a new Artist (Wedge subclass) for that as wedges are only geometric shapes and do not sore information in some absolute value. That is some work but doable.

There's a second issue in that the API changes gets larger. currently, `pie(..., label=...)` sets wedge labels and is used for legend. We want to limit it to legend. Therefore people have to move to `pie(..., wedge_labels=...)` (or `inner_labels`) that step is smaller and easier to search replace than moving to `pie(...); pie_label(...)` and also more relevant parameters with it. We could alternatively keep parameters for a single label on `pie()` (internally delegating to `pie_label()`) and only advertize `pie_label()` for further labels, but that's also inconsistent.

timhoffm commented:
> I don't like them b/c we don't force any semantic constraints on them, which means realistically you can use either one for the other case. Consistently the "why are there 3 ways to do this?" is what creates confusion.

Well, there's semantics though the defaults. Yes, you can override this by intentionally choosing other parameters. But even then, `outer_label_distance=0.7` is basically saying "I move the outer label to the inside of the wedge as well", which feels ok to me.

I would not worry too much about this. There's still one canonical way to do each use case. Yes you could do `pie(..., outer_labels=..., outer_label_distance=0.7)` to set one inner label. But let's not count this. There are various ways to intentionally write hard-to-understand code.

rcomer commented:
Just to make sure I’ve understood, if we add the `pie_label` method together with *wedge_labels*, does *wedge_labels* only do one set of labels?

It did feel a little clunky adding `wedge_label_distance=[0.6, 1.1]` through all the tests and examples 🤔

rcomer commented:
OK, but can we say the initial implementation of `PieContainer` has immutable data?  I am starting to worry about the size/scope of this...

rcomer commented:
There seems to be a consensus for the `pie_label` method.  I will add that in a new PR and this one will be "PR2".

timhoffm commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -3540,7 +3540,13 @@ def pie(self, x, explode=None, labels=None, colors=None,
+            .. deprecated:: 3.12
+                In future these labels will not appear on the wedges but only
+                be made available for the legend (see *labeldistance* below).
+                To label the wedges directly, use *wedge_labels* or the
```
Maybe
```suggestion
                To label place labels on the wedges, use *wedge_labels* or the
```

timhoffm commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -3711,11 +3755,16 @@ def get_next_color():
             pc.add_texts([])
-        else:
+
+        if labeldistance is not None:
```
Do we want to prohibit `wedge_labels is not None and labels is not None and labeldistance is not None`. This would add two sets of labels. Currently, and after deprecation, we'll only add one set of labels.

Since `wedge_labels` is new, anybody who sets that could also adapt to `labeldistance=None`; and if needed use `pie_label` to add the second set of labels.

rcomer replied:
Done.

rcomer commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -3654,6 +3687,17 @@ def pie(self, x, explode=None, labels=None, colors=None,
             fracs = x
         if labels is None:
             labels = [''] * len(x)
+        elif labeldistance is False:
```
Buying time is part of the point, since I think we agreed somewhere that functionality should not be deprecated until its alternative has been available for at least one meso version.  Also it would be strange to me that passing 1.1 gives me a warning when any other number does not.

rcomer replied:
But perhaps I should use `_api.UNSET` instead of `False`? 👀

QuLogic commented:
3.11 is branched; I don't know if you have followup corrections to make or just a rebase.

rcomer commented:
I have rebased and also changed the default _labeldistance_ from `False` to `_UNSET`.  I went through several iterations to make mypy and stubtest happy with `_UNSET` in `pyplot`.  What I've got seems to work, but I'm still not too confident I have the right thing.

rcomer commented on lib/matplotlib/_api/__init__.pyi:
```diff
@@ -19,6 +19,7 @@ from .deprecation import (  # noqa: F401, re-exported API
 _T = TypeVar("_T")
 
 class _Unset: ...
+UNSET = _Unset()
```
Without this, mypy complains `_UNSET` doesn't exist when checking `pyplot.py`.

timhoffm commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -6725,35 +6743,53 @@ def test_pie_frame_grid():
+    labels = 'Hogwarts', 'Frogs', 'Dogs', 'Logs'
+    sizes = [15, 30, 45, 10]
+    colors = ['yellowgreen', 'gold', 'lightskyblue', 'lightcoral']
+    explode = (0, 0.1, 0, 0)  # only "explode" the 2nd slice (i.e. 'Hogs')
```
Label names do not match "Hogs" is not in `labels`

rcomer replied:
I took the liberty of also fixing it in the test I copy-pasted from.

story645 commented on doc/api/next_api_changes/deprecations/29152_REC.rst:
```diff
@@ -0,0 +1,11 @@
+3.14, when it will also be deprecated and then removed in Matplotlib 3.16.  To
+preserve the existing behavior for now, set ``labeldistance=1.1``.  For the longer
+term, use the new *wedge_labels* parameter of `~.Axes.pie` or the `~.Axes.pie_label`
+method instead of *labels*.
```
Can you please add wedge_lable_distances and make  clear that pie_labels can also be used for distance?

story645 commented on doc/release/next_whats_new/pie_wedge_labels.rst:
```diff
@@ -0,0 +1,26 @@
+
+`~.Axes.pie` now accepts a *wedge_labels* parameter as a shortcut to the
+`~.Axes.pie_label` method. This may be used for simple annotation of the wedges
+of the pie chart.  It can take
+
+* a list of strings, similar to the existing *labels* parameter
+* a format string similar to the existing *autopct* parameter except that it
+  uses the `str.format` method, and it can handle absolute values as well as
+  fractions/percentages
```
```suggestion
of the pie chart.  It can take

* a list of strings, similar to the existing *labels* parameter
* a format string similar to the existing *autopct* parameter, except that it
  uses the `str.format` method and it can handle absolute values as well as
  fractions/percentages
```

story645 commented on galleries/examples/pie_and_polar_charts/pie_features.py:
```diff
@@ -15,41 +15,68 @@
 # the *wedgeprops* argument, as demonstrated in
 # :doc:`/gallery/pie_and_polar_charts/nested_pie`.
 #
+# Controlling label positions
```
```suggestion
# Set label positions
```

story645 commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -3672,8 +3707,32 @@ def pie(self, x, explode=None, labels=None, colors=None,
+            if wedge_labels is not None and labeldistance is not None:
+                raise ValueError(
+                    'wedge_labels is a replacement for labels when annotating the '
+                    'wedges, so the two should not be used together.  To add multiple'
```
```suggestion
                    'wedges, so should not be used with labeldistance.  To add multiple'
```
Might get lost which two are triggering this

rcomer replied:
I went for a slightly different wording as you could get here if you left _labeldistance_ to default.

story645 reviewed (approved):
Is all nits but also thanks for cleaning up all the examples where labels aren't shown!

rcomer commented:
I think I addressed all comments here.  Please shout if I missed something.

story645 commented:
Thanks @rcomer for seeing this through!

rcomer commented:
Thanks for all the discussion and reviews!
